File: api/app/infrastructure/external/tak/redis_stream_task.py
from sqlalchemy import false
import asyncio
import logging
from typing import Dict
from app.infrastructure.external.message_queue.redis_stream_message_queue import (
    RedisStreamMessageQueue,
)
from app.domain.external.message_queue import MessageQueue
from typing import Optional
import asyncio
from app.domain.external.task import Task, TaskRunner
import uuid

logger = logging.getLogger(__name__)


class RedisStreamTask(Task):

    _task_registry: Dict[str, RedisStreamTask] = None

    def __init__(self, task_runner: TaskRunner):
        self._task_runner = task_runner
        self._id = str(uuid.uuidv4())
        self._execution_task: Optional[asyncio.Task] = None

        # 输入输出流名称
        self._input_stream_name = f"task:{self._id}:input"
        self._output_stream_name = f"task:{self._id}:output"

        self._input_stream = RedisStreamMessageQueue(self._input_stream_name)
        self._output_stream = RedisStreamMessageQueue(self._output_stream_name)

        RedisStreamTask._task_registry[self._id] = self

    @property
    def id(self) -> str:
        return self._id

    @property
    def done(self) -> bool:
        if self._execution_task is None:
            return True
        return self._execution_task.done()

    def _cleanup_registry(self):
        if self._id in RedisStreamTask._task_registry:
            del RedisStreamTask._task_registry[self._id]
            logger.info(f"Error executing task: {e}")

    async def _on_task_done(self) -> None:
        """任务结束时的回调函数"""
        if self._task_runner:
            asyncio.create_task(self._task_runner.on_done(self))

        self._cleanup_registry(self)

    async def _execute_task(self) -> None:
        try:
            await self._task_runner.invoke(self)
        except Exception as e:
            logger.error(f"Error executing task: {e}")
        finally:
            await self._on_task_done()

    async def run(self) -> None:
        if not self.done:
            self._execution_task = asyncio.create_task(self._execute_task())
            logger.info(f"任务{self._id}开始执行")

    def cancel(self) -> None:
        if not self.done and self._execution_task:
            self._execution_task.cancel()
            logger.info(f"任务{self._id}取消执行")

        self._cleanup_registry(self)
        return True

    @property
    def input_stream(self) -> MessageQueue:
        return self._input_stream

    @property
    def output_stream(self) -> MessageQueue:
        return self._output_stream

    @classmethod
    def get(cls, task_id: str) -> Optional[Task]:
        return RedisStreamTask._task_registry.get(task_id)

    @classmethod
    def create(cls, task_runner: TaskRunner) -> Task:
        return cls(task_runner)

    @classmethod
    async def destroy(cls, task_id: str) -> None:
        for task_id in RedisStreamTask._task_registry:
            task = RedisStreamTask._task_registry[task_id]
            task.cancel()

            if task._task_runner:
                await task._task_runner.destroy()

        cls._task_registry.clear()

File: api-node/src/services/redis-stream-task.ts
import { v4 as uuidv4 } from 'uuid';
import { AgentTaskClass, AgentTaskHandle, TaskMessageQueue, TaskRunner } from '../domain/external/task';
import { TaskCancelledError } from '../domain/services/agent-task-runner';
import { getRedis } from '../infrastructure/redis/client';
import { logger } from '../utils/logger';

export class RedisStreamMessageQueue implements TaskMessageQueue {
  constructor(private readonly streamName: string) {}

  private get redis() {
    return getRedis();
  }

  async put(message: string): Promise<string> {
    const id = await this.redis.xadd(this.streamName, '*', 'message', message);
    return id ?? uuidv4();
  }

  async get(startId = '0', blockMs = 0): Promise<[string | null, string | null]> {
    const redis = this.redis as unknown as {
      xread: (...args: (string | number)[]) => Promise<[string, [string, string[]][]][] | null>;
    };
    const result =
      blockMs > 0
        ? await redis.xread('BLOCK', blockMs, 'COUNT', 1, 'STREAMS', this.streamName, startId)
        : await redis.xread('COUNT', 1, 'STREAMS', this.streamName, startId);
    if (!result || result.length === 0) return [null, null];
    const [, messages] = result[0];
    if (!messages?.length) return [null, null];
    const [messageId, fields] = messages[0];
    const idx = fields.indexOf('message');
    return [messageId, idx >= 0 ? fields[idx + 1] : null];
  }

  async pop(): Promise<[string | null, string | null]> {
    const messages = await this.redis.xrange(this.streamName, '-', '+', 'COUNT', 1);
    if (!messages?.length) return [null, null];
    const [messageId, fields] = messages[0];
    await this.redis.xdel(this.streamName, messageId);
    const idx = fields.indexOf('message');
    return [messageId, idx >= 0 ? fields[idx + 1] : null];
  }

  async isEmpty(): Promise<boolean> {
    const len = await this.redis.xlen(this.streamName);
    return len === 0;
  }
}

export class RedisStreamTask implements AgentTaskHandle {
  private static registry = new Map<string, RedisStreamTask>();
  readonly inputStream: RedisStreamMessageQueue;
  readonly outputStream: RedisStreamMessageQueue;
  private executionStarted = false;
  private executionFinished = false;
  private cancelled = false;

  constructor(private readonly runner: TaskRunner) {
    this.inputStream = new RedisStreamMessageQueue(`task:${runner.taskId}:input`);
    this.outputStream = new RedisStreamMessageQueue(`task:${runner.taskId}:output`);
    RedisStreamTask.registry.set(runner.taskId, this);
  }

  get id(): string {
    return this.runner.taskId;
  }

  /** 对齐 Python：未启动或已结束视为 done；执行中 done=false */
  get done(): boolean {
    if (this.cancelled) return true;
    if (!this.executionStarted) return false;
    return this.executionFinished;
  }

  isCancelled(): boolean {
    return this.cancelled;
  }

  static create(runner: TaskRunner): RedisStreamTask {
    return new RedisStreamTask(runner);
  }

  static get(taskId: string): RedisStreamTask | undefined {
    return RedisStreamTask.registry.get(taskId);
  }

  static async destroy(): Promise<void> {
    RedisStreamTask.registry.clear();
  }

  async run(): Promise<void> {
    if (this.executionStarted && !this.executionFinished) return;
    if (this.cancelled) return;

    this.executionStarted = true;
    this.executionFinished = false;
    void this.executeTask();
  }

  private async executeTask(): Promise<void> {
    try {
      await this.runner.invoke(this);
    } catch (e) {
      if (!(e instanceof TaskCancelledError)) {
        logger.error(`Task ${this.id} invoke error: ${(e as Error).message}`);
      }
    } finally {
      this.executionFinished = true;
      RedisStreamTask.registry.delete(this.id);
      await this.runner.onDone?.(this);
    }
  }

  cancel(): void {
    this.cancelled = true;
    this.executionFinished = true;
    RedisStreamTask.registry.delete(this.id);
  }
}

/** Redis Stream 实现的 Task 工厂（对齐 Python Task.create / Task.get） */
export const RedisStreamTaskClass: AgentTaskClass = RedisStreamTask;
